state/protocol/protocol_state/epochs/statemachine.go

```diff
@@ -237,30 +245,26 @@ func (e *EpochStateMachine) ParentState() protocol.KVStoreReader {
 // Results must be ordered by block height.
 // Expected errors during normal operations:
 // - `protocol.InvalidServiceEventError` if any service event is invalid or is not a valid state transition for the current protocol state
-func (e *EpochStateMachine) applyServiceEventsFromOrderedResults(orderedUpdates []flow.ServiceEvent) ([]transaction.DeferredDBUpdate, error) {
-	var dbUpdates []transaction.DeferredDBUpdate
+func (e *EpochStateMachine) applyServiceEventsFromOrderedResults(orderedUpdates []flow.ServiceEvent) (*transaction.DeferredBlockPersist, error) {
+	dbUpdates := transaction.NewDeferredBlockPersist()
 	for _, event := range orderedUpdates {
 		switch ev := event.Event.(type) {
 		case *flow.EpochSetup:
 			processed, err := e.activeStateMachine.ProcessEpochSetup(ev)
 			if err != nil {
-				return nil, fmt.Errorf("could not process epoch setup event: %w", err)
+				return transaction.NewDeferredBlockPersist(), fmt.Errorf("could not process epoch setup event: %w", err)
 			}
-
 			if processed {
-				// we'll insert the setup event when we insert the block
-				dbUpdates = append(dbUpdates, e.setups.StoreTx(ev))
+				dbUpdates.AddDbOp(e.setups.StoreTx(ev)) // we'll insert the setup event when we insert the block
 			}
 
 		case *flow.EpochCommit:
 			processed, err := e.activeStateMachine.ProcessEpochCommit(ev)
 			if err != nil {
-				return nil, fmt.Errorf("could not process epoch commit event: %w", err)
+				return transaction.NewDeferredBlockPersist(), fmt.Errorf("could not process epoch commit event: %w", err)
```

Comment:
```suggestion
				return nil, fmt.Errorf("could not process epoch commit event: %w", err)
```

Why not nil? Usually we return nil if there is an error.

Reply:
good point. Replaced by `nil` values in https://github.com/onflow/flow-go/pull/5644/commits/832174bff4f1ead202c9d77818a13552fa308011